index.js

```diff
@@ -47,62 +31,20 @@ module.exports = function prependFile(path, data, options) {
     throw new TypeError('Bad arguments');
   }
 
-  var appendOptions = {
-    encoding: options.encoding,
-    mode: options.mode,
-    flags: 'a'
-  };
-
-  // a temp file is written even if dist file does not exist. PR welcome for better implementation.
-  tmp
-    .file(function (err, tempFilePath, fd, cleanupCallback) {
-      if (err) {
-        callback(err);
-        return;
-      }
-
-      fs.writeFile(tempFilePath, data, options, function (err) {
-        if (err) {
-          callback(err);
-          return;
-        }
-
-        fs.createReadStream(path, options)
-          .on('error', function(err) {
-            if (err.code === 'ENOENT' /*file does not exist*/) {
-              fs.writeFile(path, data, options, function (err) {
-                if (err) {
-                  callback(err);
-                } else {
-                  callback();
-                }
-              });
-            } else {
-              callback(err);
-            }
-          })
-          .pipe(fs.createWriteStream(tempFilePath, appendOptions))
-          .on('error', function(err) {
-            callback(err);
-          })
-          .on('finish', function() {
-            fs.createReadStream(tempFilePath, options)
-              .on('error', function(err) {
-                callback(err);
-              })
-              .pipe(fs.createWriteStream(path, options))
-              .on('error', function(err) {
-                callback(err);
-              })
-              .on('finish', function() {
-                cleanupCallback();
-                callback();
-              });
-          });
-      });
-    });
+  fileExists(path)
+    .then(() => createTempFile())
+    .then(tempFile =>
+      writeToFile(tempFile.path, data, options)
+        .then(() => writeFileContentsToFile(path, tempFile.path, options, true))
```

Comment:
Not saying to avoid promises. The `.then()` is nested in another `.then()` here. You could just return the promise and `.then()` from the outer scope.